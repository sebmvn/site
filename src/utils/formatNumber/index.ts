export const formatPriceNotation = (value: string) => {
  value = value?.replace('.', '').replace(',', '').replace(/\D/g, '')

  if (isNaN(parseFloat(value))) {
    return ''
  }

  const options = { minimumFractionDigits: 2 }
  const result = new Intl.NumberFormat('pt-BR', options).format(
    parseFloat(value) / 100
  )

  return result
}

export const formatIntegerNotation = (value: string) => {
  return value?.replace('.', '').replace(',', '').replace(/\D/g, '')
}

export const toDecimal = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseFloat(formattedValue!) / 100
}

export const toInteger = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseInt(formattedValue!)
}
