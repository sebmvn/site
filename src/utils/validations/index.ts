import { CalendarType } from 'hooks/use-calc'
import Joi from 'joi'

const fieldsValidations = {
  initialValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  monthlyValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),

  interestRate: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  period: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  interestRateCalendar: Joi.string().valid('monthly', 'yearly').required(),
  periodCalendar: Joi.string().valid('monthly', 'yearly').required()
}

export type FieldErrors = {
  [key: string]: string
}

const getFieldErrors = (objError: Joi.ValidationResult) => {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type compoundInterestCalculatorFields = {
  initialValue: string
  monthlyValue: string
  interestRate: string
  period: string
  interestRateCalendar: CalendarType
  periodCalendar: CalendarType
}
export const compoundInterestCalculatorValidate = (
  values: compoundInterestCalculatorFields
) => {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
