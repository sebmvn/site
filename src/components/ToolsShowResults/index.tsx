import { Grid } from 'components/Grid'
import Heading from 'components/Heading'
import ToolsResultCard, {
  ToolsResultCardProps
} from 'components/ToolsResultCard'
import * as S from './styles'

export type ToolsShowResultsProps = {
  title?: string
  columns?: 'one' | 'two' | 'three' | 'default'
  items: ToolsResultCardProps[]
}

const ToolsShowResults = ({
  title = 'Resultados',
  columns = 'default',
  items
}: ToolsShowResultsProps) => {
  const valueOverflow = () => {
    return !!items.find((item) => item.value.length > 18)
  }

  return (
    <S.Wrapper columns={valueOverflow() ? 'one' : columns}>
      <Heading size="medium">{title}</Heading>
      <Grid>
        {items.map((item, index) => (
          <ToolsResultCard key={index} {...item} />
        ))}
      </Grid>
    </S.Wrapper>
  )
}

export default ToolsShowResults
