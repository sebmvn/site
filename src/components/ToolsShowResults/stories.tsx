import { Story, Meta } from '@storybook/react/types-6-0'
import ToolsShowResults, { ToolsShowResultsProps } from '.'

const items = [
  {
    title: 'Valor total final',
    value: 'R$ 58.877,75'
  },
  {
    title: 'Valor total investido',
    value: 'R$ 52.400,00'
  },
  {
    title: 'Total em juros',
    value: 'R$ 6.477,75'
  }
]

export default {
  title: 'ToolsShowResults',
  component: ToolsShowResults,
  args: {
    items,
    columns: 'default'
  }
} as Meta

export const Default: Story<ToolsShowResultsProps> = (args) => (
  <ToolsShowResults {...args} />
)
