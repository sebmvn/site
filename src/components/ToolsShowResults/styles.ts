import styled, { css } from 'styled-components'

import { Grid } from 'components/Grid'
import * as ToolsResultCardStyles from 'components/ToolsResultCard/styles'

type WrappersProps = {
  columns: 'one' | 'two' | 'three' | 'default'
}

const wrapperModifiers = {
  one: () => css`
    ${Grid} {
      grid-template-columns: repeat(auto-fill, minmax(100%, 1fr));
    }
  `,

  two: () => css`
    ${Grid} {
      grid-template-columns: repeat(auto-fill, minmax(50rem, 1fr));
    }
  `,

  three: () => css`
    ${Grid} {
      grid-template-columns: repeat(auto-fill, minmax(36rem, 1fr));
    }
  `,

  default: () => css`
    ${Grid} {
      grid-template-columns: repeat(auto-fill, minmax(25rem, 1fr));
    }
  `
}

export const Wrapper = styled.div<WrappersProps>`
  ${({ theme, columns }) => css`
    ${Grid} {
      ${ToolsResultCardStyles.Wrapper}:nth-child(3n+1) {
        ${ToolsResultCardStyles.Value} {
          color: ${theme.colors.gray};
        }
      }
      ${ToolsResultCardStyles.Wrapper}:nth-child(3n+2) {
        ${ToolsResultCardStyles.Value} {
          color: ${theme.colors.bannerLink};
        }
      }
      ${ToolsResultCardStyles.Wrapper}:nth-child(3n+3) {
        ${ToolsResultCardStyles.Value} {
          color: ${theme.colors.secondary};
        }
      }
    }

    ${!!columns && wrapperModifiers[columns]}
  `}
`
