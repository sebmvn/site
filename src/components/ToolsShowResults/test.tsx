import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test-utils'

import ToolsShowResults from '.'

const items = [
  {
    title: 'Valor total final',
    value: 'R$ 58.877,75'
  },
  {
    title: 'Valor total investido',
    value: 'R$ 52.400,00'
  },
  {
    title: 'Total em juros',
    value: 'R$ 6.477,75'
  }
]

describe('<ToolsShowResults />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(
      <ToolsShowResults title="ToolsShowResults" items={items} />
    )

    expect(
      screen.getByRole('heading', { name: /ToolsShowResults/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
