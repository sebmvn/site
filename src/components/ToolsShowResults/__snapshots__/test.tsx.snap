// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToolsShowResults /> should render the heading 1`] = `
.c3 {
  display: grid;
  grid-template-columns: repeat(auto-fill,minmax(25rem,1fr));
  grid-gap: 3.2rem;
  margin: 3.2rem 0;
}

.c1 {
  color: #193241;
  font-size: 1.8rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  background: #FFFFFF;
  padding: 3.2rem 2.4rem;
  border: 1px solid #F4F4F4;
  border-radius: 0.5rem;
  box-shadow: 0px 2px 4px #1932413F;
}

.c5 .c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
}

.c8 {
  color: #193241;
  font-size: 1.4rem;
  font-weight: 400;
}

.c10 {
  font-size: 2.0rem;
  font-weight: 900;
}

.c0 .c2 .c4:nth-child(3n+1) .c9 {
  color: #8696A0;
}

.c0 .c2 .c4:nth-child(3n+2) .c9 {
  color: #6E82FA;
}

.c0 .c2 .c4:nth-child(3n+3) .c9 {
  color: #0ADC7D;
}

.c0 .c2 {
  grid-template-columns: repeat(auto-fill,minmax(25rem,1fr));
}

@media (max-width:768px) {
  .c1 {
    font-size: 1.8rem;
  }
}

<div
  class="c0"
>
  <h2
    class="c1"
    color="primary"
  >
    ToolsShowResults
  </h2>
  <div
    class="c2 c3"
  >
    <article
      class="c4 c5"
    >
      <div
        class="c6 c7"
      >
        <h4
          class="c8"
        >
          Valor total final
        </h4>
        <data
          class="c9 c10"
        >
          R$ 58.877,75
        </data>
      </div>
    </article>
    <article
      class="c4 c5"
    >
      <div
        class="c6 c7"
      >
        <h4
          class="c8"
        >
          Valor total investido
        </h4>
        <data
          class="c9 c10"
        >
          R$ 52.400,00
        </data>
      </div>
    </article>
    <article
      class="c4 c5"
    >
      <div
        class="c6 c7"
      >
        <h4
          class="c8"
        >
          Total em juros
        </h4>
        <data
          class="c9 c10"
        >
          R$ 6.477,75
        </data>
      </div>
    </article>
  </div>
</div>
`;
